crates/sui-metrics-proxy/src/admin.rs

```diff
@@ -0,0 +1,163 @@
+    rustls::ServerConfig, SelfSignedCertificate, TlsAcceptor, ValidatorAllowlist,
+    ValidatorCertVerifier,
+};
+use sui_types::sui_system_state::ValidatorMetadata;
+use tokio::{signal, sync::mpsc};
+use tracing::{error, info};
+
+/// Configure our graceful shutdown scenarios
+pub async fn shutdown_signal(h: axum_server::Handle) {
+    let ctrl_c = async {
+        signal::ctrl_c()
+            .await
+            .expect("failed to install Ctrl+C handler");
+    };
+
+    #[cfg(unix)]
+    let terminate = async {
+        signal::unix::signal(signal::unix::SignalKind::terminate())
+            .expect("failed to install signal handler")
+            .recv()
+            .await;
+    };
+
+    #[cfg(not(unix))]
+    let terminate = std::future::pending::<()>();
+
+    tokio::select! {
+        _ = ctrl_c => {},
+        _ = terminate => {},
+    }
+
+    let grace = 30;
+    info!(
+        "signal received, starting graceful shutdown, grace period {} seconds, if needed",
+        &grace
+    );
+    h.graceful_shutdown(Some(Duration::from_secs(grace)))
+}
+
+/// App will configure our routes and create our mpsc channels.  This fn is also used to instrument
+/// our tests
+pub fn app(buffer_size: usize, network: String) -> Router {
+    // we accept data on our UpstreamConsumer up to our buffer size.
+    let (sender, receiver) = mpsc::channel(buffer_size);
+    let mut consumer = UpstreamConsumer::new(network, receiver);
+
+    tokio::spawn(async move { consumer.run().await });
+
+    // build our application with a route and our sender mpsc
+    Router::new()
+        .route("/publish/metrics", axum_post(publish_metrics))
+        .with_state(Arc::new(sender))
+}
+
+pub async fn server(
+    listener: std::net::TcpListener,
+    _acceptor: TlsAcceptor, // TODO enable for tls
+    app: Router,
+) -> std::io::Result<()> {
+    // setup our graceful shutdown
+    let handle = axum_server::Handle::new();
+    // Spawn a task to gracefully shutdown server.
+    tokio::spawn(shutdown_signal(handle.clone()));
+
+    axum_server::Server::from_tcp(listener)
+        // .acceptor(acceptor) // TODO enable for tls
+        .handle(handle)
+        .serve(app.into_make_service())
+        .await
+}
+
+pub fn create_server_cert(
+    hostname: &str,
+) -> Result<(ServerConfig, ValidatorAllowlist), sui_tls::rustls::Error> {
+    let mut rng = rand::thread_rng();
+    let server_keypair = Ed25519KeyPair::generate(&mut rng);
+    let server_certificate = SelfSignedCertificate::new(server_keypair.private(), hostname);
+
+    ValidatorCertVerifier::rustls_server_config(
+        vec![server_certificate.rustls_certificate()],
+        server_certificate.rustls_private_key(),
+    )
+}
+
+async fn get_validators(url: String) -> Result<Vec<ValidatorMetadata>> {
+    let client = reqwest::Client::builder().build().unwrap();
+    let request = serde_json::json!({
+        "jsonrpc": "2.0",
+        "method":"sui_getValidators",
+        "id":1,
+    });
```

Comment:
We probably need to improve our rust-sdk to be able to support calling this (I have no idea what our rust sdk can do atm fwiw). This is fine for now, just making a comment

Reply:
Yeah, i wasn't sure the best way to call, so i used the only thing i knew, `serde_json::json!`.  Magically little thing that lets you build "type safe" json, just as you would in python, et al.